pupepat/test/pupil-synth/src/pupil_plate_image.py

```diff
@@ -0,0 +1,106 @@
+import numpy as np
+from astropy.io import fits
+from src.source import Source
+
+
+class PupilPlateImage:
+    def __init__(self, shape=(4096, 4096), seed=42):
+        self.shape = shape
+        self._random = np.random.RandomState(seed)  # mine
+
+        self.sources = []
+        self._cutout_margin_factor = 1.2
+
+        # Default Noise parameters
+        self.sky_level = 200.
+        self.read_noise = self.sky_level / 100.
+        self.add_poisson_noise = True
+
+    def render(self) -> np.array:
+        """
+        Given the properties that have been set (background, sources, noise, etc)
+        generate the pupil plate image as an np.array
+
+        :return: image
+        """
+        image = np.zeros(self.shape, dtype=np.float_)
+
+        # # add the background sky level
+        image += self.sky_level
+
+        # add the sources
+        for source in self.sources:
+            # add (+=) source decal into the appropriate slice of the data
+            a = source.x - source.half_width
+            b = source.y - source.half_width
+
+            decal = source.decal
+            shape = decal.shape
+            image[a:a + shape[0], b:b + shape[1]] += decal
+
+        # add Poisson noise (shot noise)
+        if self.add_poisson_noise:
+            image = np.float_(self._random.poisson(image))
+
+        # # add read noise
+        read_noise = self._random.normal(scale=self.read_noise, size=image.size)
+        read_noise.shape = image.shape
+        image += read_noise
+
+        return image
+
+    def __str__(self) -> str:
+        return super().__str__()
+
+    def read_fits_from(self, filename):
+        self.filename = filename
+
+        # assume for now that we only read FITS file that we've written
+        # and that those have known properties
+        hdu_list = fits.open(filename)
+        assert(len(hdu_list) == 1)
```

Comment:
Can you comment on why this is necessary to have here? Why not just use fits.open?

Reply:
it's not implemented yet, but the idea is to put the information about the synthesized sources and their parameters in the FITS header. This way, a synthesized FITS file is self-contained with all it's meta-data and de-coupled from PUPE-PAT. So, after the image is synthesized, at some later date when PUPE-PAT is run on it, the "correct answers" and other info are readily available in the FITS headers.